### PR TITLE
ODC-7270, ODC-7272: Improve telemetry: Add resource and tab name to the page title

### PR DESCRIPTION
**Pull request by lokanandaprabhu**
Story- 
https://issues.redhat.com/browse/ODC-7270
https://issues.redhat.com/browse/ODC-7272

### CODE REVIEW

File: frontend/public/components/resource-list.tsx [hunk 1]
```diff
@@ -113,7 +113,9 @@ export const ResourceDetailsPage = connectToPlural((props: ResourceDetailsPagePr
   return (
     <>
       <Helmet>
-        <title>{`${decodedName} · Details`}</title>
+        <title
+          data-title-id={`${props.match.params.name} · ${props.kindObj?.kind}`}
+        >{`${decodedName} · Details`}</title>
```
Comment (jerolimov) [inline frontend/public/components/resource-list.tsx:118]: Hey, can you please:

1. Skip the resource name (pod-1234) from telemetry.
2. And can we add the translated resource kind (incl. the Helm utils you have mentioned) for the user here as well?

Thanks :)

```suggestion
        <title
          data-title-id={props.kindObj?.kind}
        >{`${decodedName} · ${translatedResourceKind} · Details`}</title>
```

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: frontend/__tests__/components/cluster-settings.spec.tsx [hunk 2]
```diff
@@ -71,9 +71,6 @@ describe('Cluster Settings page', () => {
   it('should render ClusterSettingsPage component', () => {
     expect(wrapper.exists()).toBe(true);
   });
-  it('should render correct Cluster Settings page title', () => {
-    expect(wrapper.contains('Cluster Settings')).toBeTruthy();
-  });
```
Comment (jerolimov) [inline frontend/__tests__/components/cluster-settings.spec.tsx:76]: Why did you remove this test?

File: frontend/__tests__/components/cluster-settings.spec.tsx [hunk 3]
```diff
@@ -113,20 +110,20 @@ describe('Cluster Settings page', () => {
       wrapper
         .find(HorizontalNav)
         .at(0)
-        .props().pages[0].name,
-    ).toBe('Details');
+        .props().pages[0].nameKey,
+    ).toBe('public~Details');
```
Comment (jerolimov) [inline frontend/__tests__/components/cluster-settings.spec.tsx:114]: Alternative you check this with an regex:

```suggestion
    ).toMatch('Details');
```

File: frontend/packages/console-shared/src/utils/utils.ts [hunk 4]
```diff
@@ -1,7 +1,9 @@
+import { createContext } from 'react';
 import { JSONSchema7 } from 'json-schema';
 import { toPath } from 'lodash';
+import * as _ from 'lodash';
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/utils/utils.ts:4]: You can merge this two lodash imports and import just what you need:

```suggestion
import { startCase, toPath } from 'lodash';
```

File: frontend/packages/console-telemetry-plugin/src/listeners/segment.ts [hunk 5]
```diff
@@ -106,7 +106,12 @@ export const eventListener: TelemetryEventListener = async (
       }
       break;
     case 'page':
-      (window as any).analytics.page(undefined, properties, anonymousIP);
+      {
+        const titleElement = document.querySelector('title');
+        const titleValue = titleElement.getAttribute('data-telemetry');
+        properties.title = titleValue || document.title;
+        (window as any).analytics.page(undefined, properties, anonymousIP);
+      }
```
Comment (jerolimov) [inline frontend/packages/console-telemetry-plugin/src/listeners/segment.ts:114]: Instead of applying the title property just for segment, we should do that when the 'page' event is created.

Two times in  `public/components/app.jsx`

File: frontend/packages/console-shared/src/utils/utils.ts [hunk 6]
```diff
@@ -89,3 +91,29 @@ export const alphanumericCompare = (a: string, b: string): number => {
...
+    }
+    return model.label;
+  }
+  if (translationForResourceKind[kindString]) {
+    return translationForResourceKind[kindString];
+  }
+  return _.startCase(kindString);
+};
+
+type PageTitleContextType = {
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/utils/utils.ts:114]: Export this type so that you can use it:

```suggestion
export type PageTitleContextValues = {
```

File: frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx [hunk 7]
```diff
@@ -32,6 +33,7 @@ describe('Monitoring Page ', () => {
   });
 
   it('should render all Tabs of Monitoring page for selected project', () => {
+    const { t } = useTranslation();
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx:36]: This is not a react context, you should avoid using a hook here.

File: frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx [hunk 8]
```diff
@@ -52,11 +54,12 @@ describe('Monitoring Page ', () => {
     const actualTabs = component
       .find(utils.HorizontalNav)
       .prop('pages')
-      .map((page) => page.name);
+      .map((page) => t(page.nameKey));
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx:57]: Maybe you can just remove the namespace from the i18n key?

```suggestion
      .map((page) => page.nameKey.replace('devconsole~', ''));
```

File: frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx [hunk 9]
```diff
@@ -52,11 +54,12 @@ describe('Monitoring Page ', () => {
...
     expect(actualTabs).toEqual(expectedTabs);
   });
 
   it('should not render the Alerts tab if user has no access to get prometheousRule resource', () => {
+    const { t } = useTranslation();
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx:62]: same as above

File: frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx [hunk 10]
```diff
@@ -74,7 +77,7 @@ describe('Monitoring Page ', () => {
     const actualTabs = component
       .find(utils.HorizontalNav)
       .prop('pages')
-      .map((page) => page.name);
+      .map((page) => t(page.nameKey));
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/monitoring/__tests__/MonitoringPage.spec.tsx:80]: same as above

```suggestion
      .map((page) => page.nameKey.replace('devconsole~', ''));
```

File: frontend/packages/pipelines-plugin/src/components/pipelineruns/__tests__/PipelineRunDetailsPage.spec.tsx [hunk 11]
```diff
@@ -55,15 +56,15 @@ describe('PipelineRunDetailsPage:', () => {
   it('Should contain task runs page', () => {
     const { pages } = wrapper.props();
 
-    const taskRunsPage = pages.find((page) => page.name === 'TaskRuns');
+    const taskRunsPage = pages.find((page) => page.nameKey === `${i18nPipelineNS}~TaskRuns`);
```
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/src/components/pipelineruns/__tests__/PipelineRunDetailsPage.spec.tsx:59]: Your change is okay! Alternative you check if the page name includes the right value:

```suggestion
    const taskRunsPage = pages.find((page) => page.nameKey.includes('TaskRuns');
```

File: frontend/packages/topology/src/components/list-view/list-view-utils.tsx [hunk 12]
```diff
@@ -3,11 +3,6 @@ import * as _ from 'lodash';
 import { K8sKind, modelFor } from '@console/internal/module/k8s';
 import { getResourceKind } from '../../utils/topology-utils';
 
-export const translationForResourceKind = {
-  // t('topology~Helm Release')
-  HelmRelease: `topology~Helm Release`,
-};
-
 export const labelForNodeKind = (kindString: string) => {
```
Comment (jerolimov) [inline frontend/packages/topology/src/components/list-view/list-view-utils.tsx:6]: We should not split `labelKeyForNodeKind` and `labelForNodeKind`, as you moved the labelKey function into console-shared please move this function as well.

File: frontend/public/components/factory/details.tsx [hunk 13]
```diff
@@ -147,43 +154,45 @@ export const DetailsPage = withFallback<DetailsPageProps>(({ pages = [], ...prop
...
-          breadcrumbsFor={
-            props.breadcrumbsFor ??
-            (!pluginBreadcrumbs ? breadcrumbsForDetailsPage(kindObj, props.match) : undefined)
-          }
-          resourceKeys={resourceKeys}
-          getResourceStatus={props.getResourceStatus}
-          customData={props.customData}
-          badge={props.badge || getBadgeFromType(kindObj?.badge)}
-          icon={props.icon}
+      <PageTitleContext.Provider value={titleProviderValues}>
```
Comment (jerolimov) [inline frontend/public/components/factory/details.tsx:157]: You can replace the top level React fragment (`<>`) with this `PageTitleContext.Provider` to reduce the diff here.

File: frontend/public/components/resource-list.tsx [hunk 14]
```diff
@@ -110,20 +112,24 @@ export const ResourceDetailsPage = connectToPlural((props: ResourceDetailsPagePr
...
-      <AsyncComponent
-        loader={componentLoader || defaultPage}
-        match={props.match}
-        namespace={ns}
-        kind={props.modelRef}
-        kindObj={kindObj}
-        name={decodedName}
-        badge={getBadgeFromType(kindObj.badge)}
-      />
+      <PageTitleContext.Provider value={titleProviderValues}>
```
Comment (jerolimov) [inline frontend/public/components/resource-list.tsx:122]: With just one root level node you don't need the React.Fragment (`<>`) anymore.

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 15]
```diff
@@ -206,7 +210,28 @@ export const NavBar = withRouter<NavBarProps>(({ pages, baseURL, basePath }) =>
...
+  const activePage = pages.find(({ href, path }) => {
+    const matchURL = matchPath(location.pathname, {
+      path: `${basePath}/${removeLeadingSlash(path || href)}`,
+      exact: true,
+    });
+    return matchURL?.isExact;
+  });
+
+  const labelId =
+    (activePage?.nameKey && activePage?.nameKey.split('~')[1]) || activePage?.name || 'Details';
```
Comment (jerolimov) [inline frontend/public/components/utils/horizontal-nav.tsx:222]: nit:

```suggestion
    (activePage?.nameKey?.split('~')[1]) || activePage?.name || 'Details';
```

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 16]
```diff
@@ -206,7 +210,28 @@ export const NavBar = withRouter<NavBarProps>(({ pages, baseURL, basePath }) =>
...
+  return (
+    <>
+      {titlePrefix && (
+        <Helmet>
+          <title data-telemetry={`${telemetryPrefix} · ${labelId}`}>
```
Comment (jerolimov) [inline frontend/public/components/utils/horizontal-nav.tsx:227]: In theory telemetryPrefix can be undefined (or empty), let us unsure that we don't pass it then:

```suggestion
          <title data-telemetry={telemetryPrefix ? `${telemetryPrefix} · ${labelId}` : labelId}>
```

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 17]
```diff
@@ -206,7 +210,28 @@ export const NavBar = withRouter<NavBarProps>(({ pages, baseURL, basePath }) =>
...
+      exact: true,
+    });
+    return matchURL?.isExact;
+  });
+
+  const labelId =
+    (activePage?.nameKey && activePage?.nameKey.split('~')[1]) || activePage?.name || 'Details';
+  return (
+    <>
+      {titlePrefix && (
```
Comment (jerolimov) [inline frontend/public/components/utils/horizontal-nav.tsx:225]: Is there a reason to not set the current activePage.name as title when there is no `titlePrefix`?

Couldn't we set this always. But you're right we should expect that the `titlePrefix is not defined.

File: frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json [hunk 18]
```diff
@@ -91,9 +91,7 @@
   "No task runs found": "No task runs found",
   "Download all task logs": "Download all task logs",
   "No logs found": "No logs found",
-  "PipelineRun logs": "PipelineRun logs",
   "Triggered by": "Triggered by",
-  "PipelineRun events": "PipelineRun events",
```
Comment (jerolimov) [inline frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json:96]: This surprised me. We should check the pipelines pages that everything is still set correct.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Thanks for this big change. From the code perspective and this looks almost good to me. And some comments anyway and I also need to test the telemetry behavior!

**Comment by jerolimov** [comment]
The title in the dom is updated really well. But I found some cases where it doesn't work. Can you take a look why the modal could not resolved in this cases:

In the admin perspective > Compute > **_Nodes_** > Select a node works fine

image

But when I select Compute > **_Machines_** > Select a machine I see the "internal modal string" as `telemetryPrefix`:

image

File: frontend/__tests__/components/cluster-settings.spec.tsx (hunk 2 above)
Comment (lokanandaprabhu) [reply frontend/__tests__/components/cluster-settings.spec.tsx:76]: Hi @jerolimov ,

It is directly rendering ClusterSettingsPage component and it had <title> component with 'Cluster Settings' text but now I removed <Helmet> in ClusterSettingsPage so the wrapper will not have 'Cluster Settings' text.

File: frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json (hunk 18 above)
Comment (lokanandaprabhu) [reply frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json:98]: Before we had separate <Helmet> components for PipelineRun logs and PipelineRun events pages, now these are not required

File: frontend/packages/console-telemetry-plugin/src/listeners/segment.ts (hunk 5 above)
Comment (lokanandaprabhu) [reply frontend/packages/console-telemetry-plugin/src/listeners/segment.ts:114]: In Segment, we are checking the event type and based the event type we are making network call. I added above code for eventType is page. Basically all the page events will be captured here.

**Comment by lokanandaprabhu** [comment]
Hi @jerolimov ,

I updated the PR based the PR comments and added comments for couple of queries you asked above. PTAL.

File: frontend/packages/console-telemetry-plugin/src/listeners/segment.ts (hunk 5 above)
Comment (jerolimov) [reply frontend/packages/console-telemetry-plugin/src/listeners/segment.ts:114]: But other telemetry listeners might be also interested in the same page title. This should be a general telemetry improvement, not just for the segment implementation.

**Comment by jerolimov** [comment]
Any chance to add some unit or e2e to test that the document.title is changed when switching between tabs on some multi list pages (like pipelines) and detail pages (details, yaml, etc.)? Or do we have this already?

File: frontend/packages/console-telemetry-plugin/src/listeners/segment.ts (hunk 5 above)
Comment (lokanandaprabhu) [reply frontend/packages/console-telemetry-plugin/src/listeners/segment.ts]: Added in app.jsx

**Comment by jerolimov** [comment]
/retitle ODC-7270, ODC-7272: Improve telemetry: Add resource and tab name to the page title

File: frontend/__tests__/components/cluster-settings.spec.tsx (hunk 2 above)
Comment (jerolimov) [reply frontend/__tests__/components/cluster-settings.spec.tsx:76]: Ok

File: frontend/public/components/cluster-settings/cluster-settings.tsx [hunk 19]
```diff
@@ -1353,29 +1353,35 @@ export const ClusterSettingsPage: React.FC<ClusterSettingsPageProps> = ({ match
...
+      nameKey: 'public~Configuration',
       component: GlobalConfigPage,
     },
   ];
+  const titleProviderValues = {
+    telemetryPrefix: 'Cluster Settings',
+    titlePrefix: title,
+  };
   return (
     <>
```
Comment (jerolimov) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:1378]: This React.Fragment could be removed when there is just one component

File: frontend/packages/console-shared/src/utils/utils.ts [hunk 20]
```diff
@@ -89,3 +90,37 @@ export const alphanumericCompare = (a: string, b: string): number => {
...
+  }
+  return startCase(kindString);
+};
+
+export type PageTitleContextValues = {
+  telemetryPrefix?: string;
+  titlePrefix?: string;
+};
+
+export const PageTitleContext = createContext<PageTitleContextValues>({});
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/utils/utils.ts:126]: Can you move this info `packages/console-shared/src/components/pagetitle/PageTitleContext.tsx` or `packages/console-shared/src/context/pagetitle/PageTitleContext.tsx`?

I would use context, but other contexts are part of the components folder, see: `packages/console-shared/src/components/toast/ToastContext.tsx`, so that's also fine, but I prefer to not anything new to `src/utils/utils.ts`. You can also move the utils to `packages/console-shared/src/components/pagetitle/utils.tsx` or at least a `page-title-utils.ts` here in the utils folder.

Wdyt?

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 21]
```diff
@@ -206,7 +210,27 @@ export const NavBar = withRouter<NavBarProps>(({ pages, baseURL, basePath }) =>
...
+  });
+
+  const labelId = activePage?.nameKey?.split('~')[1] || activePage?.name || 'Details';
+  return (
+    <>
+      <Helmet>
+        <title data-telemetry={telemetryPrefix ? `${telemetryPrefix} · ${labelId}` : labelId}>
+          {titlePrefix
+            ? `${titlePrefix} · ${t(activePage?.nameKey) || activePage?.name}`
+            : `${t(activePage?.nameKey) || activePage?.name}`}
```
Comment (jerolimov) [inline frontend/public/components/utils/horizontal-nav.tsx:228]: Normally we don't call t() when nameKey is null or undefined. But it's working fine and doesn't log a warning/error, so it's okayish for me. But I would prefer:

```suggestion
            ? `${titlePrefix} · ${activePage?.nameKey ? t(activePage.nameKey) : activePage?.name}`
            : `${activePage?.nameKey ? t(activePage.nameKey) : activePage?.name}`}
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
As mentioned in slack, I got many errors logged that some strings couldn't be translated on the Helm > Repositories and Pipelines > Repositories details page

image

Failed when calculate `titlePrefix` in this (compiled) code:

```

    const titleProviderValues = {
        telemetryPrefix: (_c = props === null || props === void 0 ? void 0 : props.kindObj) === null || _c === void 0 ? void 0 : _c.kind,
        titlePrefix: `${props.name} · ${t(Object(_console_shared__WEBPACK_IMPORTED_MODULE_3__["labelKeyForNodeKind"])((_d = props === null || props === void 0 ? void 0 : props.kindObj) === null || _d === void 0 ? void 0 : _d.kind))}`,
    };
```

File: frontend/public/components/utils/telemetry.ts [hunk 22]
```diff
@@ -19,3 +20,21 @@ export const withoutSensitiveInformations = (location: Location): Location => {
...
+    const last = titleArray[titleArray.length - 1];
+    if (titleArray.length === 1) {
+      return pageTitleValue;
+    } else if (titleArray.length && last === productName) {
+      const newTelemetrytitle = pageTitleValue.replace(` · ${productName}`, '');
+      return newTelemetrytitle;
+    }
+  }
+  return titleValue || document.title;
+};
```
Comment (jerolimov) [inline frontend/public/components/utils/telemetry.ts:40]: It works, but when I need to check the details and different paths here I want recommend a IMHO simpler solution. Wdyt about this:

```suggestion

const titleProductNameSuffix = ` · ${getBrandingDetails().productName}`;

export const getTelemetryTitle = () => {
  const titleElement = document.querySelector('title');
  let title = titleElement.getAttribute('data-telemetry') || document.title;
  if (title.endsWith(titleProductNameSuffix)) {
    title = title.substring(0, title.length - titleProductNameSuffix.length);
  }
  return title;
};
```

File: frontend/packages/topology/src/filters/KindFilterDropdown.tsx [hunk 23]
```diff
@@ -80,7 +80,7 @@ const KindFilterDropdown: React.FC<KindFilterDropdownProps> = ({
           data-test={filter.label}
         >
           <ResourceIcon kind={filter.id} />
-          {t(filter.labelKey)} ({supportedKinds[filter.id]})
+          {filter.labelKey} ({supportedKinds[filter.id]})
```
Comment (jerolimov) [inline frontend/packages/topology/src/filters/KindFilterDropdown.tsx:83]: Let us improve this in a follow-up PR. This might confuse other developers in the future.

**Review by jerolimov — APPROVED** [review]
As mentioned I have three comments:

1. In a follow up PR we should check if we can rename `labelKeyForNodeKind` because it returns now a translated label and no labelKey anymore!
2. Helm > Repositories and Pipelines > Repositories return both "Repositories" as telemetry title. The user sees "Pipeline Repositories" and "Helm Repositories".
   * I think we should keep such a prefix also for the telemetry to differentiate both pages.
   * We should do it similarly for all multi list pages, esp. the Pipeline pages in the Admin perspective.
   * We can handle that as a follow-up story / PR.

/lgtm
/approve

File: frontend/packages/console-shared/src/components/multi-tab-list/MultiTabListPage.tsx [hunk 24]
```diff
@@ -19,6 +20,7 @@ interface MultiTabListPageProps {
   pages: Page[];
   match: Rmatch<any>;
   secondaryButtonAction?: SecondaryButtonAction;
+  navMenu?: string;
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/components/multi-tab-list/MultiTabListPage.tsx:23]: The navigation and a page are not connected, or not connected in this direction. This props sounds to me, oh I can add a (sub) navigation configuration the MultiTabList, but that's not the case here.

Recommended names: `titleKey` or `telemetryPrefix`

Both makes it more clear that this will reflect the (displayed) page title more then the navigation title:

image

image

File: frontend/packages/knative-plugin/src/components/overview/serving-list/ServingListsPage.tsx [hunk 25]
```diff
@@ -67,6 +70,7 @@ const ServingListPage: React.FC<ServingListPageProps> = ({ match }) => {
         match={match}
         title={t('knative-plugin~Serving')}
         menuActions={menuActions}
+        navMenu="Serving"
```
Comment (jerolimov) [inline frontend/packages/knative-plugin/src/components/overview/serving-list/ServingListsPage.tsx:73]: Not essential to get this in, but I saw that the Serverless > Serving and Serverless Evening pages in the admin perspective use this for telemetry but not in the UI. I think that's fine, as it is like before. :+1:

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/components/multi-tab-list/MultiTabListPage.tsx (hunk 24 above)
Comment (lokanandaprabhu) [reply frontend/packages/console-shared/src/components/multi-tab-list/MultiTabListPage.tsx]: Changed the name to `telemetryPrefix`

**Comment by sanketpathak** [comment]
Verified on cluster
Screenshot from 2023-04-05 03-25-55

/label qe-approved

**Review by vikram-raj — COMMENTED** [review]
/lgtm

/retest
